File: macos/Tests/ElevateCoreTests/ActivationProfileTests.swift
import Testing
import Foundation
@testable import ElevateCore

@Suite struct ActivationProfileTests {
    func key(_ n: String, kind: RoleScopeKind = .entraDirectory, tenantId: String = "t") -> RoleKey {
        switch kind {
        case .entraDirectory: RoleKey(identityId: "i", tenantId: tenantId, scope: .entraDirectory(roleDefinitionId: n, directoryScopeId: "/"))
        case .azureResource: RoleKey(identityId: "i", tenantId: tenantId, scope: .azureResource(scope: "/subscriptions/s", roleDefinitionId: n))
        case .group: RoleKey(identityId: "i", tenantId: tenantId, scope: .group(groupId: n, accessId: .member))
        }
    }

    @Test func stateWithoutProfilesDecodes() throws {
        let json = #"{"identities":[],"tenants":[],"manualRoles":[],"memory":[]}"#
        let s = try JSONDecoder().decode(AppState.self, from: Data(json.utf8))
        #expect(s.profiles.isEmpty)
        let minimal = try JSONDecoder().decode(AppState.self, from: Data("{}".utf8))
        #expect(minimal.identities.isEmpty && minimal.profiles.isEmpty)
    }

    @Test func profilesRoundTripAndHelpers() throws {
        var s = AppState()
        let p = ActivationProfile(name: "Ops", entries: [.init(roleKey: key("a"), lastDuration: .seconds(3600))], lastJustification: "INC")
        s.upsertProfile(p)
        s.upsertProfile(ActivationProfile(name: "Second", entries: []))
        let decoded = try JSONDecoder().decode(AppState.self, from: JSONEncoder().encode(s))
        #expect(decoded.profiles.count == 2)
        #expect(decoded.profile(id: p.id)?.entries.first?.lastDuration == .seconds(3600))
        var renamed = p; renamed.name = "Ops 2"
        s.upsertProfile(renamed)
        #expect(s.profiles.count == 2 && s.profile(id: p.id)?.name == "Ops 2")
        s.moveProfile(fromOffsets: IndexSet(integer: 1), toOffset: 0)
        #expect(s.profiles.first?.name == "Second")
        s.removeProfile(id: p.id)
        #expect(s.profiles.count == 1)
    }

    @Test func removingTenantDropsProfileEntries() throws {
        var s = AppState()
        let keepKey = key("keep", tenantId: "t1")
        let dropKey = key("drop", tenantId: "t2")
        let p = ActivationProfile(name: "Mixed", entries: [.init(roleKey: keepKey), .init(roleKey: dropKey)])
        s.upsertProfile(p)
        s.removeTenant(TenantKey(identityId: "i", tenantId: "t2"))
        let entries = s.profile(id: p.id)?.entries ?? []
        #expect(entries.count == 1)
        #expect(entries.first?.roleKey == keepKey)
    }

    @Test func removingIdentityDropsProfileEntries() throws {
        var s = AppState()
        let keepKey = RoleKey(identityId: "other", tenantId: "t1", scope: .entraDirectory(roleDefinitionId: "keep", directoryScopeId: "/"))
        let dropKey = key("drop", tenantId: "t1")
        let p = ActivationProfile(name: "Mixed", entries: [.init(roleKey: keepKey), .init(roleKey: dropKey)])
        s.upsertProfile(p)
        s.removeIdentity("i")
        let entries = s.profile(id: p.id)?.entries ?? []
        #expect(entries.count == 1)
        #expect(entries.first?.roleKey == keepKey)
    }

    @Test func pinnedDecodesTolerantlyAndIsWrittenOnlyWhenTrue() throws {
        let legacy = #"{"id":"A1B2C3D4-E5F6-4A5B-8C7D-9E0FABCDEF12","name":"Ops","entries":[]}"#
        let decoded = try JSONDecoder().decode(ActivationProfile.self, from: Data(legacy.utf8))
        #expect(decoded.pinned == false)
        let unpinnedJSON = String(decoding: try JSONEncoder().encode(decoded), as: UTF8.self)
        #expect(!unpinnedJSON.contains("pinned"))
        var pinned = decoded; pinned.pinned = true
        let pinnedJSON = String(decoding: try JSONEncoder().encode(pinned), as: UTF8.self)
        #expect(pinnedJSON.contains(#""pinned":true"#))
        #expect(try JSONDecoder().decode(ActivationProfile.self, from: Data(pinnedJSON.utf8)).pinned)
    }

    @Test func pinningIsCappedAtTheLimit() {
        var s = AppState()
        let profiles = (0...ProfilePins.limit).map { ActivationProfile(name: "P\($0)", entries: []) }
        for p in profiles { s.upsertProfile(p) }
        // `#expect` captures its operands immutably, so each mutating call lands in a local first.
        for p in profiles.prefix(ProfilePins.limit) {
            let pinned = s.setPinned(id: p.id, true)
            #expect(pinned)
        }
        #expect(s.pinnedProfiles.count == ProfilePins.limit)
        let fifth = s.setPinned(id: profiles[ProfilePins.limit].id, true)
        #expect(!fifth)
        #expect(s.profile(id: profiles[ProfilePins.limit].id)?.pinned == false)
        // Re-pinning an already pinned profile is not a new pin.
        let repinned = s.setPinned(id: profiles[0].id, true)
        #expect(repinned)
        let unpinned = s.setPinned(id: profiles[0].id, false)
        #expect(unpinned)
        #expect(s.pinnedProfiles.map(\.name) == ["P1", "P2", "P3"])
        let afterUnpin = s.setPinned(id: profiles[ProfilePins.limit].id, true)
        #expect(afterUnpin)
        let unknown = s.setPinned(id: UUID(), true)
        #expect(!unknown)
    }

    @Test func summaryCaption() {
        #expect(ProfileSummary.caption(entries: [.init(roleKey: key("a"), lastDuration: nil)]) == "1 role")
        #expect(ProfileSummary.caption(entries: [.init(roleKey: key("a"), lastDuration: nil), .init(roleKey: key("b", kind: .azureResource), lastDuration: nil)]) == "2 roles")
        #expect(ProfileSummary.caption(entries: [.init(roleKey: key("g", kind: .group), lastDuration: nil)]) == "1 group")
        #expect(ProfileSummary.caption(entries: [.init(roleKey: key("a"), lastDuration: nil), .init(roleKey: key("g", kind: .group), lastDuration: nil), .init(roleKey: key("h", kind: .group), lastDuration: nil)]) == "1 role · 2 groups")
        #expect(ProfileSummary.caption(entries: []) == "empty")
    }

    @Test func sourceDefaultsToUserAndIsWrittenOnlyWhenManaged() throws {
        let user = ActivationProfile(name: "Ops", entries: [])
        #expect(user.source == .user)
        let userJSON = String(decoding: try JSONEncoder().encode(user), as: UTF8.self)
        #expect(!userJSON.contains("source"))

        var managed = user
        managed.source = .managed
        #expect(managed != user)
        let managedJSON = String(decoding: try JSONEncoder().encode(managed), as: UTF8.self)
        #expect(managedJSON.contains("\"source\":\"managed\""))
        let decoded = try JSONDecoder().decode(ActivationProfile.self, from: Data(managedJSON.utf8))
        #expect(decoded.source == .managed)

        // A profile written before `source` existed decodes as a user profile.
        let old = #"{"id":"A1B2C3D4-E5F6-4A5B-8C7D-9E0FABCDEF12","name":"Old","entries":[]}"#
        #expect(try JSONDecoder().decode(ActivationProfile.self, from: Data(old.utf8)).source == .user)
    }
}

File: windows/tests/Elevate.Core.Tests/ActivationProfileTests.cs
using Elevate.Core.Models;
using Elevate.Core.Storage;
using FluentAssertions;

namespace Elevate.Core.Tests;

public class ActivationProfileTests
{
    private static RoleKey Key(string name, RoleScopeKind kind = RoleScopeKind.EntraDirectory, string tenantId = "t") =>
        new("i", tenantId, kind switch
        {
            RoleScopeKind.EntraDirectory => new EntraDirectoryScope(name, "/"),
            RoleScopeKind.AzureResource => new AzureResourceScope("/subscriptions/s", name),
            _ => new GroupScope(name, GroupAccess.Member),
        });

    [Fact]
    public void StateWithoutProfilesDecodes()
    {
        const string json = """{"identities":[],"tenants":[],"manualRoles":[],"memory":[]}""";

        Json.Deserialize<AppState>(json)!.Profiles.Should().BeEmpty();

        var minimal = Json.Deserialize<AppState>("{}")!;
        minimal.Identities.Should().BeEmpty();
        minimal.Profiles.Should().BeEmpty();
    }

    [Fact]
    public void ProfilesRoundTripAndHelpers()
    {
        var state = new AppState();
        var profile = new ActivationProfile("Ops", [new ActivationProfile.Entry(Key("a"), TimeSpan.FromHours(1))], "INC");
        state.UpsertProfile(profile);
        state.UpsertProfile(new ActivationProfile("Second", []));

        var decoded = Json.Deserialize<AppState>(Json.Serialize(state))!;

        decoded.Profiles.Should().HaveCount(2);
        decoded.Profile(profile.Id)!.Entries[0].LastDuration.Should().Be(TimeSpan.FromHours(1));
        decoded.Profile(profile.Id)!.LastJustification.Should().Be("INC");

        state.UpsertProfile(profile with { Name = "Ops 2" });
        state.Profiles.Should().HaveCount(2);
        state.Profile(profile.Id)!.Name.Should().Be("Ops 2");

        state.MoveProfiles([1], 0);
        state.Profiles[0].Name.Should().Be("Second");

        state.RemoveProfile(profile.Id);
        state.Profiles.Should().ContainSingle();
    }

    [Fact]
    public void ProfileIdIsWrittenAsSwiftsUppercaseUuid()
    {
        var id = Guid.Parse("a1b2c3d4-e5f6-4a5b-8c7d-9e0fabcdef12");
        var profile = new ActivationProfile(id, "Ops", []);

        var json = Json.Serialize(profile);

        json.Should().Be("""{"id":"A1B2C3D4-E5F6-4A5B-8C7D-9E0FABCDEF12","name":"Ops","entries":[]}""");
        Json.Deserialize<ActivationProfile>(json)!.Id.Should().Be(id);
    }

    [Fact]
    public void RemovingTenantDropsProfileEntries()
    {
        var state = new AppState();
        var keep = Key("keep", tenantId: "t1");
        var drop = Key("drop", tenantId: "t2");
        var profile = new ActivationProfile("Mixed", [new ActivationProfile.Entry(keep), new ActivationProfile.Entry(drop)]);
        state.UpsertProfile(profile);

        state.RemoveTenant(new TenantKey("i", "t2"));

        state.Profile(profile.Id)!.Entries.Select(e => e.RoleKey).Should().Equal(keep);
    }

    [Fact]
    public void RemovingIdentityDropsProfileEntries()
    {
        var state = new AppState();
        var keep = new RoleKey("other", "t1", new EntraDirectoryScope("keep", "/"));
        var drop = Key("drop", tenantId: "t1");
        var profile = new ActivationProfile("Mixed", [new ActivationProfile.Entry(keep), new ActivationProfile.Entry(drop)]);
        state.UpsertProfile(profile);

        state.RemoveIdentity("i");

        state.Profile(profile.Id)!.Entries.Select(e => e.RoleKey).Should().Equal(keep);
    }

    [Fact]
    public void PinnedDecodesTolerantlyAndIsWrittenOnlyWhenTrue()
    {
        const string legacy = """{"id":"A1B2C3D4-E5F6-4A5B-8C7D-9E0FABCDEF12","name":"Ops","entries":[]}""";

        var decoded = Json.Deserialize<ActivationProfile>(legacy)!;

        decoded.Pinned.Should().BeFalse();
        Json.Serialize(decoded).Should().NotContain("pinned");

        decoded.Pinned = true;
        var json = Json.Serialize(decoded);

        json.Should().Contain("\"pinned\":true");
        Json.Deserialize<ActivationProfile>(json)!.Pinned.Should().BeTrue();
        decoded.DeepCopy().Pinned.Should().BeTrue();
        decoded.Should().NotBe(Json.Deserialize<ActivationProfile>(legacy));
    }

    [Fact]
    public void PinningIsCappedAtTheLimit()
    {
        var state = new AppState();
        var profiles = Enumerable.Range(0, ProfilePins.Limit + 1).Select(i => new ActivationProfile($"P{i}", [])).ToList();
        profiles.ForEach(state.UpsertProfile);

        foreach (var p in profiles.Take(ProfilePins.Limit))
        {
            state.SetPinned(p.Id, true).Should().BeTrue();
        }

        state.PinnedProfiles.Should().HaveCount(ProfilePins.Limit);
        state.SetPinned(profiles[ProfilePins.Limit].Id, true).Should().BeFalse();
        state.Profile(profiles[ProfilePins.Limit].Id)!.Pinned.Should().BeFalse();

        // Re-pinning an already pinned profile is not a new pin.
        state.SetPinned(profiles[0].Id, true).Should().BeTrue();
        state.SetPinned(profiles[0].Id, false).Should().BeTrue();
        state.PinnedProfiles.Select(p => p.Name).Should().Equal("P1", "P2", "P3");
        state.SetPinned(profiles[ProfilePins.Limit].Id, true).Should().BeTrue();
        state.SetPinned(Guid.NewGuid(), true).Should().BeFalse();
    }

    [Fact]
    public void SummaryCaption()
    {
        ProfileSummary.Caption([new ActivationProfile.Entry(Key("a"))]).Should().Be("1 role");
        ProfileSummary.Caption([
            new ActivationProfile.Entry(Key("a")),
            new ActivationProfile.Entry(Key("b", RoleScopeKind.AzureResource)),
        ]).Should().Be("2 roles");
        ProfileSummary.Caption([new ActivationProfile.Entry(Key("g", RoleScopeKind.Group))]).Should().Be("1 group");
        ProfileSummary.Caption([
            new ActivationProfile.Entry(Key("a")),
            new ActivationProfile.Entry(Key("g", RoleScopeKind.Group)),
            new ActivationProfile.Entry(Key("h", RoleScopeKind.Group)),
        ]).Should().Be("1 role · 2 groups");
        ProfileSummary.Caption([]).Should().Be("empty");
    }

    [Fact]
    public void SourceDefaultsToUserAndIsWrittenOnlyWhenManaged()
    {
        var user = new ActivationProfile("Ops", []);
        user.Source.Should().Be(ProfileSource.User);
        Json.Serialize(user).Should().NotContain("source");

        var managed = user.DeepCopy() with { Source = ProfileSource.Managed };
        managed.Should().NotBe(user);   // source takes part in equality
        var json = Json.Serialize(managed);
        json.Should().Contain("\"source\":\"managed\"");
        Json.Deserialize<ActivationProfile>(json)!.Source.Should().Be(ProfileSource.Managed);

        // A profile written before `source` existed decodes as a user profile.
        const string Old = """{"id":"A1B2C3D4-E5F6-4A5B-8C7D-9E0FABCDEF12","name":"Old","entries":[]}""";
        Json.Deserialize<ActivationProfile>(Old)!.Source.Should().Be(ProfileSource.User);
    }
}
